pkg/collector/clock.go

```diff
@@ -140,18 +184,41 @@ func (c *fakeClock) Step(d time.Duration) {
 		// Collect timer functions to run and remove them from list.
 		newTimers := make([]*fakeTimer, 0, len(c.timers))
 		for _, t := range c.timers {
-			if !t.targetTime.After(c.now) {
-				timerFuncs = append(timerFuncs, t.f)
-			} else {
+			if t.targetTime.After(c.now) {
 				newTimers = append(newTimers, t)
+			} else {
+				expiredTimers = append(expiredTimers, t)
 			}
 		}
 		c.timers = newTimers
 	}()
-	// Run the timer functions, without holding a lock. This allows these functions to call
-	// clock.Now(), but also timer.Stop().
-	for _, f := range timerFuncs {
-		f()
+	for _, t := range expiredTimers {
+		if t.f != nil {
+			// Run the timer function, without holding a lock. This allows these
+			// functions to call clock.Now(), but also timer.Stop().
+			t.f()
+		} else {
+			retry := func() bool {
+				if !c.m.TryRLock() {
+					return true
+				}
+				defer c.m.RUnlock()
+				// If timer has been stopped or reset, do not fire.
+				if t.targetTime.IsZero() || t.targetTime.After(c.now) {
+					return false
+				}
+				select {
+				case t.ch <- c.now:
```

Comment:
I think t.ch is used to check if a timer is fired? Then why don't we need it after executing a function at line 199?
The other question is timer.C() won't return when a time is stopped?

Reply:
There is no `t.ch` for function-based timers (created with `AfterFunc`).

> The other question is timer.C() won't return when a time is stopped?

From the `time` documentation:

> For a chan-based timer created with NewTimer(d), as of Go 1.23, any receive from t.C after Stop has returned is guaranteed to block rather than receive a stale time value from before the Stop; if the program has not received from t.C already and the timer is running, Stop is guaranteed to return true. 

So it is correct, the channel will block if the timer has been stopped.

Reply:
Thanks for the explanation